File: lib/core/models/medicine.dart
class Medicine {
  final String id;
  final String userId;
  final String? pregnancyId;
  final String? babyId;
  final String name;
  final String dosage;
  final String frequency;
  final int timesPerDay;
  final DateTime startDate;
  final DateTime? endDate;
  final bool isActive;
  final String? notes;
  final DateTime createdAt;

  Medicine({
    required this.id,
    required this.userId,
    this.pregnancyId,
    this.babyId,
    required this.name,
    required this.dosage,
    required this.frequency,
    required this.timesPerDay,
    required this.startDate,
    this.endDate,
    required this.isActive,
    this.notes,
    required this.createdAt,
  });

  factory Medicine.fromMap(Map<String, dynamic> map) {
    return Medicine(
      id: map['id'] as String,
      userId: map['user_id'] as String,
      pregnancyId: map['pregnancy_id'] as String?,
      babyId: map['baby_id'] as String?,
      name: map['name'] as String,
      dosage: map['dosage'] as String,
      frequency: map['frequency'] as String,
      timesPerDay: map['times_per_day'] as int,
      startDate: DateTime.parse(map['start_date'] as String),
      endDate: map['end_date'] != null ? DateTime.parse(map['end_date'] as String) : null,
      isActive: (map['is_active'] as int) == 1,
      notes: map['notes'] as String?,
      createdAt: DateTime.parse(map['created_at'] as String),
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'user_id': userId,
      'pregnancy_id': pregnancyId,
      'baby_id': babyId,
      'name': name,
      'dosage': dosage,
      'frequency': frequency,
      'times_per_day': timesPerDay,
      'start_date': startDate.toIso8601String(),
      'end_date': endDate?.toIso8601String(),
      'is_active': isActive ? 1 : 0,
      'notes': notes,
      'created_at': createdAt.toIso8601String(),
    };
  }
}

File: backend/app/schemas/medicine.py
"""
Medicine Pydantic schemas for request/response validation.

These schemas define the structure of medicine data sent to and received from the API.
"""

from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime


class MedicineBase(BaseModel):
    """Base medicine schema with common fields."""
    user_id: Optional[str] = None
    baby_id: Optional[str] = None
    medicine_name: str = Field(..., min_length=1, max_length=255)
    dosage: str
    type: Optional[str] = None
    frequency_type: str
    frequency_value: Optional[int] = None
    scheduled_times: Optional[List[str]] = None
    start_date: str  # ISO date format YYYY-MM-DD
    end_date: Optional[str] = None  # ISO date format YYYY-MM-DD
    medicine_image_url: Optional[str] = None
    is_active: int = 1


class MedicineCreate(MedicineBase):
    """Schema for creating a new medicine."""
    id: Optional[str] = None  # Allow client to provide ID
    created_at: Optional[str] = None  # Allow client to provide timestamp


class MedicineUpdate(BaseModel):
    """Schema for updating a medicine."""
    medicine_name: Optional[str] = Field(None, min_length=1, max_length=255)
    dosage: Optional[str] = None
    type: Optional[str] = None
    frequency_type: Optional[str] = None
    frequency_value: Optional[int] = None
    scheduled_times: Optional[List[str]] = None
    start_date: Optional[str] = None
    end_date: Optional[str] = None
    medicine_image_url: Optional[str] = None
    is_active: Optional[int] = None


class MedicineResponse(MedicineBase):
    """
    Schema for medicine response.
    Used in GET /medicines endpoints.
    """
    id: str
    created_at: str

    class Config:
        from_attributes = True


class MedicineLogBase(BaseModel):
    """Base medicine log schema."""
    medicine_id: str
    user_id: str
    logged_date: str  # ISO date format YYYY-MM-DD
    logged_at: str  # ISO datetime format
    status: str  # Status: taken, missed, or skipped
    notes: Optional[str] = None


class MedicineLogCreate(MedicineLogBase):
    """Schema for creating a medicine log."""
    id: str  # Client-generated UUID


class MedicineLogResponse(MedicineLogBase):
    """Schema for medicine log response."""
    id: str

    class Config:
        from_attributes = True
